### PR TITLE
Ordered folder playback (#83)

### PR DESCRIPTION
**Pull request by viktordick**

### CODE REVIEW

**Comment by viktordick** [comment]
Attempts to fix #83

**Comment by viktordick** [comment]
Idea for refactoring - not sure if it is a good idea to include this here or if it should be handled separately:

- Change the "if the playback of this object required an override, switch on recursion for this subpath even if by default we were not trying to recurse" to "if the playback of this object requires an override, rename the old object to a new ID that is not used yet, create the new object and move the subobjects over before deleting the old object"
- Change `playback` so it always handles exactly one object, returning some information, and does not recurse
- in `playback_paths`, build a list of paths that are to be played back. Iterate by picking the first item from the list, playing it back and - if recursing - prepending all the file system contents of the folder.
- From the returned information from `playback`, check if the meta information has a field `contents`. If so, append the info to a list. Once the original list is depleted, iterate backwards through this built list and call a method that fixes the ordering of the contents according to this list.
- `write_after_recurse_hook` can be replaced by a function that only handles this case and only is defined for ordered folders, since it is only used for this.

**Review by jan-jockusch — COMMENTED** [review]
For the most part, this is a much welcome code cleanup.

The core functionality change in `playback` and `playback_paths` has not yet made its way into my brain. Everything up to those changes LGTM!

File: perfact/zodbsync/zodbsync.py [hunk 1]
```diff
@@ -704,14 +690,18 @@ def playback_paths(self, paths, recurse=True, override=False,
...
                     path=path,
                     override=override,
                     recurse=recurse,
                     skip_errors=skip_errors,
                     encoding=encoding,
-                )
+                ))
+            for result in reversed(results):
+                if result:
+                    self.playback_after_recurse(**result)
```
Comment (jan-jockusch) [inline perfact/zodbsync/zodbsync.py:704]: Oh, so doing the hook in reverse order is what fixes the Ordered Folder playback issue?

**Review by jan-jockusch — COMMENTED** [review]
Do I get it right that the only functional change up until now is to reverse the order in which the after_playback hook is executed?

File: perfact/zodbsync/zodbsync.py (hunk 1 above)
Comment (viktordick) [reply perfact/zodbsync/zodbsync.py:704]: Correct, see my comment in #83 - the ordering of an ordered folder was adjusted before its (new) children were even present, if they were created in the same operation, so they were not respected.

Not sure if this fixes all issues with Ordered Folders, but this was the issue that I could reproduce.

The refactoring I would like to see I started in #86, but did not finish yet. I will continue that when I get to it, but this bugfix should come first.

**Review by jan-jockusch — APPROVED** [review]
LGTM!

File: perfact/zodbsync/zodbsync.py (hunk 1 above)
Comment (viktordick) [reply perfact/zodbsync/zodbsync.py:704]: Slight correction: it is not that the hook is executed in reverse order. It is that the hook for the parent is executed only after all its children are created. So the `reversed` could it principle also be removed, as long the hook is executed in a separate loop.

But in the recursion case the hook was executed on the "way up", so to keep it consistent I included the `reversed`.

**Pull request merged**